Add MVP version of the library (#1)
Wikia/AlphaB: README.md

# AlphaB

## Python library for rendering charts and computing statistics for A/B testing

### Current state

This library is in the very initial state. Currently, it supports only A/B testings (two groups).

### AlphaB allows you to:

* Automatically generate charts from A/B testings
* Compute statistics in order to confirm a statistical significance between groups

> <!-- <p align="center"><img src="images/AlphaB.png" width="250"/></p> -->

Reviewer: Is it some kind of placeholder for a logo or you forgot to remove it?

    )
    bucket_test.render()
    bucket_test.compute_pvalues()
```

### Arguments

When creating a bucket test, you can specify the following arguments:

* `df` - data frame to be used for the bucket test. It is recommended to group the data frame before passing it (e.g.: When doing a bucket test on the group `design`, you should group the data frame by design and date first)
* `variable` - specifies the values on the y-axis for the chart and statistical significance check
* `group` - the name of the column which the data frame is grouped by
* `x_axis` (default: `date`) - specifies the values on the x_axis for the chart
* `custom_title` (default: `{variable} per {group}`) - specifies the title for the chart

For the `render()` method, those options can be specified to customize your chart:

* `figure_size_x` (default: 12) - the width of the chart (in inches)
* `figure_size_y` (default: 5) - the height of the chart (in inches)
* `line_width` (default: 3) - the line width in a line chart (in points)
* `title_font_size` (default: 16) -  the font size of the title in the figure 
* `legend_font_size` (default: 14) -  the font size of the legend in the figure

> In the `compute_pvalues()` method, you can customize the p-value used to reject a null hypothesis by adjusting `alpha` value (default: 0.01).

Reviewer 2: ```suggestion
In the `compute_pvalues()` method, you can customize the p-value used to reject a null hypothesis by adjusting the `alpha` value (default: `0.01`).
```

> Recommended values are: 0.01, 0.05, 0.1.

Reviewer 2: ```suggestion
Recommended values are: `0.01`, `0.05`, `0.1`.
```

Read more about statistical significance and p-value [here](https://www.statsdirect.com/help/basics/p_values.htm).

> I also, recommend [this research paper](http://www.scielo.br/pdf/bpsr/v7n1/02.pdf) for those who wants better understand those topics.

Reviewer 2: ```suggestion
[This research paper](http://www.scielo.br/pdf/bpsr/v7n1/02.pdf) is also a good place to start for those who want to better understand those topics.
```

### Screenshots

A generated chart and statistical significance analysis example:

<p align="center">
  <img width="100%" src="example/example_impressions_by_group.png" />
</p>

## Next steps

* Customize the number of groups that are taken into account A/B/C testings, A/B/C/D testings, A/B/C/D/E testings
* Render charts and compute p-values for data from more than one data frame
* Create tests for `render` and `compute_pvalue` methods
* Handle other `x_axis` that date only
* Customize names of images `plt.savefig(Path(""))`

## How to contribute

You can contribute by forking this repository, looking through the [issues](https://github.com/Wikia/AlphaB/issues) of the repository, and opening a PR on your fork. Please make sure to write a clear PR description and to provide examples for how your new feature works.

## Contributors

> * The method for checking statistical significance was highly inspired by the work of **Paulina Gralak [@Loczi94](https://github.com/Loczi94)**.

Reviewer: ❤️

* [Numpy](https://numpy.org/) - a library providing support for large, multi-dimensional arrays and matrices, along with a large collection of high-level mathematical functions
* [Scipy](https://www.scipy.org/) - a library used for scientific computing and technical computing
* [Pathlib](https://docs.python.org/3/library/pathlib.html) - offers a set of classes to handle filesystem paths

```python
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from scipy import stats
from pathlib import Path
```

## How to use it

It is highly recommended to use [Jupyter](https://jupyter.org/) to perform A/B testing analysis in Python, and AlphaB is built to be used in Jupyter Notebooks. 

Here is an example usage for AlphaB (this example doesn't include specifying a data set for now):

```python
#!/usr/bin/env python3

from bucket_test import BucketTest, GroupBy
import pandas as pd

def main():
    df = pd.DataFrame()
    bucket_test = BucketTest(
        df=df[df['device'] == 'mobile'],
        y_axis='impressions',
        group='design',
        x_axis='date',
        custom_title='Impressions by design for mobile',
        custom_ylabel='#',

>         custom_interval=1

Reviewer 2: This option should have a more explicit name 🤔 

WDYT of `custom_day_interval`? And what happens if the `x_axis` isn't a date?

Author: Thank you! I will change `custom_interval` with `custom_day_interval`.
Currently, I would like to focus on the shipping of this and for most cases - the x-axis would be a date in A/B testings in general. Will keep it in mind and will add this to `Next steps` in the README.md.

    )
    bucket_test.render()
    bucket_test.compute_pvalues()
```

### Arguments

> * ```df``` - assigns previously grouped data frame to this argument; this can include specification of a certain value within a column in the data frame (exp. `df[df['device' == 'mobile']]`)

Reviewer 2: It's not clear here what is a `previously grouped data frame`. WDYT of

```suggestion
* `df` - data frame to be used for the bucket test. It is recommended to group the data frame before passing it (e.g.: When doing a bucket test on the group `design`, you should group the data frame by design and date first)
```

Also, triple back ticks should be used for multi-line code blocks, while single back ticks are for in-line code blocks, which should be the case here :)

Author: Changed, thank you!

> * ```y_axis``` - specifies the values on the y_axis for the chart and statistical significance check

Reviewer 2: Actually, what do you think of renaming this `y_axis` to `variable`? From what I've seen, in A/B testing, what we called the `y_axis` here is usually called the variable 🤔 But I might be wrong!

```suggestion
* `y_axis` - the name of the column to use on the Y axis of the chart and statistical significance check.
```

> * ```group``` - assigns a column in the data frame which contains groups for A/B testing

Reviewer 2: WDYT of

```suggestion
* `group` - the name of the column which the data frame is grouped by
```

> * ```x_axis``` (default: `date`) - specifies the values on the x_axis for the chart

Reviewer 2: ```suggestion
* `x_axis` (default: `date`) - the name of the column to use on the X axis of the chart
```

> * ```custom_title``` (default: `y_axis`) - specifies the title for the chart

Reviewer 2: The default value for custom title isn't `y_axis`, it's `{y_axis} per {group}`

```suggestion
* `custom_title` (default: `{y_axis} per {group}`) - a custom title for the chart
```

> 

Reviewer 2: Unnecessary newline

```suggestion
```

### Screenshots

A generated chart and statistical significance analysis example:

<p align="center">
  <img width="100%" src="example/example_impressions_by_group.png" />
</p>

## Next steps

* Customize the number of groups that are taken into account A/B/C testings, A/B/C/D testings, A/B/C/D/E testings.

> * Group data frame so that it can be given to `bucket_test.render()`

Reviewer 2: What do you mean by that?

Author: As discussed, I will remove it, since this shouldn't be in the scope of the library. Thank you for giving me tips on this type of stuff! :)

> * Render charts and compute p-values for data from more than one data frame

Reviewer 2: Shouldn't that be done by doing multiple bucket tests? It sounds to me like one bucket test should do one rendering and one computing of pvalues, if there are multiple tests to handle, they should be in separate tests, otherwise the library would become overcomplicated and it definitely shouldn't :)

Author: Sure, will consider that! Thanks!

* Create tests for `render` and `compute_pvalue` methods

## How to contribute

> You can contribute by creating a new branch on the repository and a PR 

Reviewer 2: I'd suggest something maybe a little less vague than this :p WDYT of

```suggestion
You can contribute by forking this repository, looking through the [issues](https://github.com/Wikia/AlphaB/issues) of the repository, and opening a PR on your fork. Please make sure to write a clear PR description and to provide examples for how your new feature works.
```

Maybe something like that?

## Contributors

The method for checking statistical significance was highly inspired by the work of **Paulina Gralak [@Loczi94](https://github.com/Loczi94)**.

> Thank a lot!

Reviewer 2: ```suggestion
Thanks a lot!
```
